portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java

```diff
@@ -0,0 +1,117 @@
+      response.setContentType("text/html");
+      response.getWriter().println("Please upload a single file.");
+      return;
+    }
+
+    // Get input from the form.
+    String text = getParameter(request, TEXT, /* DefaultValue= */ "");
+    String commenterName = getParameter(request, NAME, /* DefaultValue= */ "Anonymous");
+    String commenterEmail = getParameter(request, EMAIL, /* DefaultValue= */ "Unknown");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(DatastoreHelper.buildCommentEntity(
+        commenterEmail, commenterName, imageUrl, text, System.currentTimeMillis()));
+
+    // Redirect to same HTML page.
+    response.sendRedirect("/index.html");
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private static String getUploadedFileUrl(HttpServletRequest request, String formInputElementName)
+      throws Exception {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(IMAGE);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // User submitted more than 1 file, so we throw an exception.
+    if (blobKeys.size() > 1) {
+      throw new Exception(String.format("Number of files %d uploaded exceeds 1.", blobKeys.size()));
+    }
+
+    // Get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
+  }
+
+  /*
+   * Converts List of comments into a JSON using the gson library.
+   */
+  private static String convertToJson(List<Comment> comments) {
+    return new Gson().toJson(comments);
+  }
+
+  private static String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    return (value == null || value.isEmpty()) ? defaultValue : value;
+  }
+}
```

Comment:
These can be in a CommonUtils.java file, and shared between servlet.

Reply:
Done.